R/cov_struct.R

```diff
@@ -0,0 +1,442 @@
+#'
+#' @param cov (`cov_struct`)\cr
+#'   A covariance structure object
+#'
+#' @return A string used for method dispatch when passed to TMB
+#'
+#' @keywords internal
+tmb_cov_type <- function(cov) {
+  paste0(cov$type, if (cov$heterogeneous) "h")
+}
+
+
+
+#' Define a Covariance Structure
+#'
+#' @param type (`string`)\cr
+#'   The name of the covariance structure type to use. For available options,
+#'   see `cov_types()`. If a type abbreviation is used that implies
+#'   heterogenicity (e.g. `cph`) and no value is provided to `heterogeneous`,
+#'   then the heterogenicity is derived from the type name.
+#' @param visits (`character`)\cr
+#'   A vector of variable names to use for the longitudinal terms of the
+#'   covariance structure. Multiple terms are only permitted for the `"spatial"`
+#'   covariance type.
+#' @param subject (`string`)\cr
+#'   The name of the variable that encodes a subject identifier.
+#' @param group (`string`)\cr
+#'   Optionally, the name of the variable that encodes a grouping variable for
+#'   subjects.
+#'
+#' @return A `cov_struct` object
+#'
+#' @examples
+#' cov_struct("csh", "AVISITN", "USUBJID")
+#' cov_struct("spatial", c("VISITA", "VISITB"), group = "GRP", subject = "SBJ")
+#'
+#' @family cov_struct
+#' @export
+cov_struct <- function(type = cov_types(), visits, subject, group = character(),
+  heterogeneous = FALSE) {
+
+  # if heterogeneous isn't provided, derive from provided type
+  if (missing(heterogeneous)) {
+    heterogeneous <- switch(type,
+      toeph = , ar1h = , adh = , csh = TRUE,
+      heterogeneous
+    )
+  }
+
+  # coerce all type options into abbreviated form
+  type <- match.arg(type)
+  type <- cov_type_abbr(type)
+
+  x <- structure(
+    list(
+      type = type,
+      heterogeneous = heterogeneous,
+      visits = visits,
+      subject = subject,
+      group = group
+    ),
+    class = c("cov_struct", "mmrm_cov_struct", "list")
+  )
+
+  validate_cov_struct(x)
+}
+
+
+
+#' Validate Covariance Structure Data
+#'
+#' Run checks against relational integrity of covariance definition
+#'
+#' @param x (`cov_struct`)\cr
+#'   A covariance structure object
+#'
+#' @return `x` if successful, or an error is thrown otherwise
+#'
+#' @importFrom checkmate makeAssertCollection
+#' @keywords internal
+validate_cov_struct <- function(x) {
+  checks <- checkmate::makeAssertCollection()
+
+  with(x, {
+    assert_character(subject, len = 1, add = checks)
+    assert_logical(heterogeneous, len = 1, add = checks)
+
+    if (length(group) > 1 || length(visits) < 1) {
+      checks$push(
+        "Covariance structure must be of the form `time | (group /) subject`"
+      )
+    }
+
+    if (!type %in% cov_types(filter = "spatial") && length(visits) > 1) {
+      checks$push(paste(
+        "Non-spatial covariance structures must have a single longitudinal",
+        "variable"
+      ))
+    }
+  })
+
+  with_error_call(reportAssertions(checks), sys.call(-1))
```

Comment:
this is nice. however I would still like to keep the current assertion structure for now, and have a discussion first whether we want to go into this direction. can you please take that code out of this PR and move on a separate branch/ draft PR?